File: ruche-connectee/mobile-app/lib/services/ruche_service.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:ruche_connectee/services/firebase_realtime_service.dart';
import 'package:ruche_connectee/services/logger_service.dart';

class RucheService {
  final FirebaseRealtimeService _firebaseService;

  RucheService(this._firebaseService);

  // Collections Realtime Database pour les ruches
  static const String _collectionRuches = 'ruches';
  static const String _collectionRuchers = 'ruchers';

  /// Ajoute une nouvelle ruche dans Firebase Realtime Database
  ///
  /// Paramètres :
  /// - [idRucher] : ID du rucher auquel appartient la ruche
  /// - [nom] : nom de la ruche
  /// - [position] : position de la ruche dans le rucher
  /// - [enService] : statut de service (par défaut true)
  /// - [dateInstallation] : date d'installation (optionnel, utilise la date actuelle par défaut)
  ///
  /// Retourne l'ID du document créé
  ///
  /// Lève une exception si l'utilisateur n'est pas connecté ou si le rucher n'existe pas
  Future<String> ajouterRuche({
    required String idRucher,
    required String nom,
    required String position,
    bool enService = true,
    DateTime? dateInstallation,
  }) async {
    try {
      LoggerService.info(
          '🐝 Tentative d\'ajout d\'une nouvelle ruche: $nom dans le rucher: $idRucher');

      // Vérifier que l'utilisateur est connecté
      final User? currentUser = _firebaseService.currentUser;
      if (currentUser == null) {
        LoggerService.error(
            'Tentative d\'ajout de ruche sans utilisateur connecté');
        throw Exception(
            'Utilisateur non connecté. Veuillez vous connecter pour ajouter une ruche.');
      }

      LoggerService.debug('Utilisateur connecté: ${currentUser.uid}');

      // Vérifier que le rucher existe et appartient à l'utilisateur
      final Map<String, dynamic>? rucherDoc = await _firebaseService
          .getDocument(_collectionRuchers, idRucher);

      if (rucherDoc == null) {
        LoggerService.error('Rucher non trouvé: $idRucher');
        throw Exception('Le rucher spécifié n\'existe pas.');
      }

      // Vérifier que le rucher appartient à l'utilisateur connecté
      if (rucherDoc['idApiculteur'] != currentUser.uid) {
        LoggerService.error(
            'Tentative d\'ajout de ruche dans un rucher non autorisé: $idRucher');
        throw Exception(
            'Vous n\'êtes pas autorisé à ajouter une ruche dans ce rucher.');
      }

      // Vérifier que le rucher est actif
      if (rucherDoc['actif'] != true) {
        LoggerService.error(
            'Tentative d\'ajout de ruche dans un rucher inactif: $idRucher');
        throw Exception(
            'Impossible d\'ajouter une ruche dans un rucher inactif.');
      }

      LoggerService.debug('Rucher validé pour l\'ajout de ruche');

      // Préparer les données de la ruche
      final Map<String, dynamic> rucheData = {
        'idRucher': idRucher,
        'nom': nom.trim(),
        'position': position.trim(),
        'enService': enService,
        'dateInstallation': dateInstallation != null
            ? dateInstallation.millisecondsSinceEpoch
            : DateTime.now().millisecondsSinceEpoch,
        'dateCreation': DateTime.now().millisecondsSinceEpoch,
        // Champs additionnels pour la compatibilité
        'actif': true,
        'idApiculteur': currentUser.uid, // Pour faciliter les requêtes
      };

      LoggerService.debug('Données de la ruche à créer: $rucheData');

      // Générer un ID unique pour la ruche
      final String rucheId = DateTime.now().millisecondsSinceEpoch.toString();

      // Ajouter la ruche
      await _firebaseService.setDocument(_collectionRuches, rucheId, rucheData);

      // Mettre à jour le nombre de ruches dans le rucher
      final int nombreRuchesActuel = rucherDoc['nombreRuches'] ?? 0;
      await _firebaseService.updateDocument(_collectionRuchers, idRucher, {
        'nombreRuches': nombreRuchesActuel + 1,
        'dateModification': DateTime.now().millisecondsSinceEpoch,
      });

      LoggerService.info('🐝 Ruche créée avec succès. ID: $rucheId');

      return rucheId;
    } catch (e) {
      LoggerService.error('Erreur lors de l\'ajout de la ruche', e);

      // Gestion spécifique des erreurs Firebase
      if (e is Exception) {
        rethrow;
      }

      // Erreur générique
      throw Exception(
          'Une erreur inattendue s\'est produite lors de l\'ajout de la ruche: $e');
    }
  }

  /// Récupère toutes les ruches d'un rucher spécifique, triées par nom croissant
  ///
  /// Paramètres :
  /// - [idRucher] : ID du rucher
  ///
  /// Retourne une liste des ruches triées par nom croissant
  Future<List<Map<String, dynamic>>> obtenirRuchesParRucherTrieesParNom(
      String idRucher) async {
    try {
      // Vérifier que l'utilisateur est connecté
      final User? currentUser = _firebaseService.currentUser;
      if (currentUser == null) {
        LoggerService.error(
            'Tentative de récupération des ruches sans utilisateur connecté');
        throw Exception(
            'Utilisateur non connecté. Veuillez vous connecter pour accéder aux ruches.');
      }

      LoggerService.info(
          '🐝 Récupération des ruches pour le rucher: $idRucher (triées par nom)');

      // Vérifier que le rucher existe et appartient à l'utilisateur
      final Map<String, dynamic>? rucherDoc = await _firebaseService
          .getDocument(_collectionRuchers, idRucher);

      if (rucherDoc == null) {
        throw Exception('Le rucher spécifié n\'existe pas.');
      }

      if (rucherDoc['idApiculteur'] != currentUser.uid) {
        throw Exception(
            'Vous n\'êtes pas autorisé à accéder aux ruches de ce rucher.');
      }

      // Récupérer toutes les ruches et filtrer par rucher
      final List<Map<String, dynamic>> toutesRuches = await _firebaseService
          .getAllDocuments(_collectionRuches);

      final List<Map<String, dynamic>> ruches = toutesRuches
          .where((ruche) =>
              ruche['idRucher'] == idRucher && ruche['actif'] == true)
          .toList();

      // Trier par nom croissant (insensible à la casse)
      ruches.sort((a, b) {
        final nomA = (a['nom'] as String?)?.toLowerCase() ?? '';
        final nomB = (b['nom'] as String?)?.toLowerCase() ?? '';
        return nomA.compareTo(nomB);
      });

      LoggerService.info(
          '🐝 ${ruches.length} ruche(s) récupérée(s) avec succès pour le rucher: $idRucher (triées par nom)');

      return ruches;
    } catch (e) {
      LoggerService.error(
          'Erreur lors de la récupération des ruches triées par nom', e);
      rethrow;
    }
  }

  /// Récupère toutes les ruches d'un rucher spécifique
  ///
  /// Paramètres :
  /// - [idRucher] : ID du rucher
  ///
  /// Retourne une liste des ruches triées par nom croissant (nouvelle version)
  Future<List<Map<String, dynamic>>> obtenirRuchesParRucher(
      String idRucher) async {
    // Utiliser la nouvelle méthode avec tri par nom
    return obtenirRuchesParRucherTrieesParNom(idRucher);
  }

  /// Récupère toutes les ruches de l'utilisateur connecté
  ///
  /// Retourne une liste de toutes les ruches appartenant à l'utilisateur
  Future<List<Map<String, dynamic>>> obtenirRuchesUtilisateur() async {
    try {
      // Vérifier que l'utilisateur est connecté
      final User? currentUser = _firebaseService.currentUser;
      if (currentUser == null) {
        throw Exception('Utilisateur non connecté');
      }

      LoggerService.info(
          'Récupération de toutes les ruches pour l\'utilisateur: ${currentUser.uid}');

      final List<Map<String, dynamic>> toutesRuches = await _firebaseService
          .getAllDocuments(_collectionRuches);

      final List<Map<String, dynamic>> ruches = toutesRuches
          .where((ruche) =>
              ruche['idApiculteur'] == currentUser.uid && ruche['actif'] == true)
          .toList();

      // Trier par date de création décroissante
      ruches.sort((a, b) {
        final dateA = a['dateCreation'] as int? ?? 0;
        final dateB = b['dateCreation'] as int? ?? 0;
        return dateB.compareTo(dateA);
      });

      LoggerService.info(
          '${ruches.length} ruche(s) trouvée(s) pour l\'utilisateur');

      return ruches;
    } catch (e) {
      LoggerService.error(
          'Erreur lors de la récupération des ruches utilisateur', e);
      rethrow;
    }
  }

  /// Récupère une ruche spécifique par son ID
  ///
  /// Paramètres :
  /// - [rucheId] : ID de la ruche
  ///
  /// Retourne les données de la ruche ou null si non trouvée
  Future<Map<String, dynamic>?> obtenirRucheParId(String rucheId) async {
    try {
      // Vérifier que l'utilisateur est connecté
      final User? currentUser = _firebaseService.currentUser;
      if (currentUser == null) {
        throw Exception('Utilisateur non connecté');
      }

      LoggerService.info('Récupération de la ruche: $rucheId');

      final Map<String, dynamic>? data = await _firebaseService
          .getDocument(_collectionRuches, rucheId);

      if (data == null) {
        LoggerService.warning('Ruche non trouvée: $rucheId');
        return null;
      }

      // Vérifier que la ruche appartient à l'utilisateur connecté
      if (data['idApiculteur'] != currentUser.uid) {
        LoggerService.warning(
            'Tentative d\'accès à une ruche non autorisée: $rucheId');
        throw Exception('Accès non autorisé à cette ruche');
      }

      data['id'] = rucheId;

      LoggerService.info('Ruche récupérée avec succès: $rucheId');

      return data;
    } catch (e) {
      LoggerService.error('Erreur lors de la récupération de la ruche', e);
      rethrow;
    }
  }

  /// Met à jour une ruche existante
  ///
  /// Paramètres :
  /// - [rucheId] : ID de la ruche à mettre à jour
  /// - [nom] : nouveau nom de la ruche (optionnel)
  /// - [position] : nouvelle position de la ruche (optionnel)
  /// - [enService] : nouveau statut de service (optionnel)
  /// - [dateInstallation] : nouvelle date d'installation (optionnel)
  Future<void> mettreAJourRuche({
    required String rucheId,
    String? nom,
    String? position,
    bool? enService,
    DateTime? dateInstallation,
  }) async {
    try {
      // Vérifier que l'utilisateur est connecté
      final User? currentUser = _firebaseService.currentUser;
      if (currentUser == null) {
        throw Exception('Utilisateur non connecté');
      }

      LoggerService.info('Mise à jour de la ruche: $rucheId');

      // Vérifier que la ruche existe et appartient à l'utilisateur
      final rucheExistante = await obtenirRucheParId(rucheId);
      if (rucheExistante == null) {
        throw Exception('Ruche non trouvée');
      }

      // Préparer les données de mise à jour
      final Map<String, dynamic> updateData = {
        'dateModification': DateTime.now().millisecondsSinceEpoch,
      };

      if (nom != null) updateData['nom'] = nom.trim();
      if (position != null) updateData['position'] = position.trim();
      if (enService != null) updateData['enService'] = enService;
      if (dateInstallation != null) {
        updateData['dateInstallation'] = dateInstallation.millisecondsSinceEpoch;
      }

      // Mettre à jour le document
      await _firebaseService.updateDocument(
          _collectionRuches, rucheId, updateData);

      LoggerService.info('Ruche mise à jour avec succès: $rucheId');
    } catch (e) {
      LoggerService.error('Erreur lors de la mise à jour de la ruche', e);
      rethrow;
    }
  }

  /// Supprime une ruche (suppression logique)
  ///
  /// Paramètres :
  /// - [rucheId] : ID de la ruche à supprimer
  Future<void> supprimerRuche(String rucheId) async {
    try {
      // Vérifier que l'utilisateur est connecté
      final User? currentUser = _firebaseService.currentUser;
      if (currentUser == null) {
        throw Exception('Utilisateur non connecté');
      }

      LoggerService.info('Suppression de la ruche: $rucheId');

      // Vérifier que la ruche existe et appartient à l'utilisateur
      final rucheExistante = await obtenirRucheParId(rucheId);
      if (rucheExistante == null) {
        throw Exception('Ruche non trouvée');
      }

      final String idRucher = rucheExistante['idRucher'];

      // Suppression logique de la ruche
      await _firebaseService.updateDocument(_collectionRuches, rucheId, {
        'actif': false,
        'dateSuppression': DateTime.now().millisecondsSinceEpoch,
      });

      // Décrémenter le nombre de ruches dans le rucher
      final Map<String, dynamic>? rucherDoc = await _firebaseService
          .getDocument(_collectionRuchers, idRucher);
      
      if (rucherDoc != null) {
        final int nombreRuchesActuel = rucherDoc['nombreRuches'] ?? 0;
        await _firebaseService.updateDocument(_collectionRuchers, idRucher, {
          'nombreRuches': nombreRuchesActuel - 1,
          'dateModification': DateTime.now().millisecondsSinceEpoch,
        });
      }

      LoggerService.info('Ruche supprimée avec succès: $rucheId');
    } catch (e) {
      LoggerService.error('Erreur lors de la suppression de la ruche', e);
      rethrow;
    }
  }

  /// Stream pour écouter les changements des ruches d'un rucher en temps réel
  ///
  /// Paramètres :
  /// - [idRucher] : ID du rucher
  ///
  /// Retourne un stream de la liste des ruches triées par nom croissant
  Stream<List<Map<String, dynamic>>> ecouterRuchesParRucher(String idRucher) {
    final User? currentUser = _firebaseService.currentUser;
    if (currentUser == null) {
      LoggerService.error(
          'Tentative d\'écoute des ruches sans utilisateur connecté');
      return Stream.error(Exception(
          'Utilisateur non connecté. Veuillez vous connecter pour écouter les ruches.'));
    }

    LoggerService.info(
        '🐝 Démarrage de l\'écoute temps réel des ruches pour le rucher: $idRucher (triées par nom)');

    return _firebaseService
        .watchCollection(_collectionRuches)
        .map((toutesRuches) {
      final ruches = toutesRuches
          .where((ruche) =>
              ruche['idRucher'] == idRucher && ruche['actif'] == true)
          .toList();

      // Trier par nom croissant (insensible à la casse)
      ruches.sort((a, b) {
        final nomA = (a['nom'] as String?)?.toLowerCase() ?? '';
        final nomB = (b['nom'] as String?)?.toLowerCase() ?? '';
        return nomA.compareTo(nomB);
      });

      LoggerService.debug(
          '🐝 Mise à jour temps réel: ${ruches.length} ruche(s) pour le rucher: $idRucher (triées par nom)');

      return ruches;
    }).handleError((error) {
      LoggerService.error('Erreur dans l\'écoute temps réel des ruches', error);
      throw error;
    });
  }

  /// Stream pour écouter les changements de toutes les ruches de l'utilisateur
  ///
  /// Retourne un stream de la liste de toutes les ruches de l'utilisateur
  Stream<List<Map<String, dynamic>>> ecouterRuchesUtilisateur() {
    final User? currentUser = _firebaseService.currentUser;
    if (currentUser == null) {
      return Stream.error(Exception('Utilisateur non connecté'));
    }

    LoggerService.info(
        'Démarrage de l\'écoute temps réel de toutes les ruches utilisateur');

    return _firebaseService
        .watchCollection(_collectionRuches)
        .map((toutesRuches) {
      final ruches = toutesRuches
          .where((ruche) =>
              ruche['idApiculteur'] == currentUser.uid && ruche['actif'] == true)
          .toList();

      // Trier par date de création décroissante
      ruches.sort((a, b) {
        final dateA = a['dateCreation'] as int? ?? 0;
        final dateB = b['dateCreation'] as int? ?? 0;
        return dateB.compareTo(dateA);
      });

      LoggerService.debug(
          'Mise à jour temps réel: ${ruches.length} ruche(s) utilisateur');

      return ruches;
    });
  }
}

File: src/services/rucheService.ts
import { 
  collection, 
  addDoc, 
  getDocs, 
  doc, 
  getDoc, 
  updateDoc, 
  deleteDoc, 
  query, 
  where, 
  orderBy, 
  onSnapshot,
  Timestamp,
  runTransaction,
  WriteBatch,
  writeBatch
} from 'firebase/firestore';
import { auth, db } from '../firebase-config';
import { DonneesCapteursService, DonneesCapteur } from './donneesCapteursService';

export interface Ruche {
  id?: string;
  idRucher: string;
  nom: string;
  position: string;
  enService: boolean;
  dateInstallation: Date;
  dateCreation?: Date;
  dateMiseAJour?: Date;
  idApiculteur: string;
  actif?: boolean;
  // Champs optionnels pour les extensions futures
  nombreCadres?: number;
  typeRuche?: string;
  notes?: string;
}

export interface RucheAvecRucher extends Ruche {
  rucherNom?: string;
  rucherAdresse?: string;
}



export class RucheService {
  private static readonly COLLECTION_NAME = 'ruches';
  private static readonly RUCHERS_COLLECTION = 'ruchers';

  /**
   * Ajoute une nouvelle ruche avec validation complète
   */
  static async ajouterRuche(rucheData: {
    idRucher: string;
    nom: string;
    position: string;
    enService?: boolean;
    dateInstallation?: Date;
  }): Promise<string> {
    const currentUser = auth.currentUser;
    if (!currentUser) {
      throw new Error('Aucun utilisateur connecté. Veuillez vous connecter pour ajouter une ruche.');
    }

    console.log('Ajout d\'une ruche pour l\'utilisateur:', currentUser.uid);

    try {
      // Utiliser une transaction pour assurer la cohérence
      const result = await runTransaction(db, async (transaction) => {
        // 1. Vérifier que le rucher existe et appartient à l'utilisateur
        const rucherRef = doc(db, this.RUCHERS_COLLECTION, rucheData.idRucher);
        const rucherDoc = await transaction.get(rucherRef);

        if (!rucherDoc.exists()) {
          throw new Error('Le rucher spécifié n\'existe pas.');
        }

        const rucherData = rucherDoc.data();
        if (rucherData.idApiculteur !== currentUser.uid) {
          throw new Error('Vous n\'avez pas les permissions pour ajouter une ruche à ce rucher.');
        }

        if (!rucherData.actif) {
          throw new Error('Impossible d\'ajouter une ruche à un rucher inactif.');
        }

        // 2. Vérifier l'unicité de la position dans le rucher
        const q = query(
          collection(db, this.COLLECTION_NAME),
          where('idRucher', '==', rucheData.idRucher),
          where('position', '==', rucheData.position.trim()),
          where('actif', '==', true)
        );
        const existingRuches = await getDocs(q);

        if (!existingRuches.empty) {
          throw new Error(`Une ruche existe déjà à la position "${rucheData.position}" dans ce rucher.`);
        }

        // 3. Vérifier l'unicité du nom dans le rucher
        const qNom = query(
          collection(db, this.COLLECTION_NAME),
          where('idRucher', '==', rucheData.idRucher),
          where('nom', '==', rucheData.nom.trim()),
          where('actif', '==', true)
        );
        const existingNoms = await getDocs(qNom);

        if (!existingNoms.empty) {
          throw new Error(`Une ruche avec le nom "${rucheData.nom}" existe déjà dans ce rucher.`);
        }

        // 4. Créer la nouvelle ruche
        const nouvelleRuche: Omit<Ruche, 'id'> = {
          idRucher: rucheData.idRucher,
          nom: rucheData.nom.trim(),
          position: rucheData.position.trim(),
          enService: rucheData.enService ?? true,
          dateInstallation: rucheData.dateInstallation || new Date(),
          dateCreation: new Date(),
          dateMiseAJour: new Date(),
          idApiculteur: currentUser.uid,
          actif: true
        };

        // 5. Ajouter la ruche
        const rucheRef = await addDoc(collection(db, this.COLLECTION_NAME), {
          ...nouvelleRuche,
          dateInstallation: Timestamp.fromDate(nouvelleRuche.dateInstallation),
          dateCreation: Timestamp.now(),
          dateMiseAJour: Timestamp.now()
        });

        // 6. Mettre à jour le nombre de ruches dans le rucher
        const currentNombreRuches = rucherData.nombreRuches || 0;
        transaction.update(rucherRef, {
          nombreRuches: currentNombreRuches + 1
        });

        return rucheRef.id;
      });

      console.log('Ruche ajoutée avec succès, ID:', result);
      return result;

    } catch (error: any) {
      console.error('Erreur lors de l\'ajout de la ruche:', error);
      
      if (error.code === 'permission-denied') {
        throw new Error('Permissions insuffisantes pour ajouter une ruche.');
      } else if (error.code === 'unavailable') {
        throw new Error('Service Firestore temporairement indisponible. Veuillez réessayer.');
      } else if (error.message) {
        throw error; // Relancer les erreurs de validation personnalisées
      } else {
        throw new Error('Impossible d\'ajouter la ruche. Vérifiez votre connexion internet.');
      }
    }
  }

  /**
   * Récupère toutes les ruches d'un rucher
   */
  static async obtenirRuchesParRucher(idRucher: string): Promise<Ruche[]> {
    try {
      const currentUser = auth.currentUser;
      if (!currentUser) {
        throw new Error('Utilisateur non connecté');
      }

      const q = query(
        collection(db, this.COLLECTION_NAME),
        where('idRucher', '==', idRucher),
        where('idApiculteur', '==', currentUser.uid),
        where('actif', '==', true),
        orderBy('position')
      );

      const querySnapshot = await getDocs(q);
      const ruches: Ruche[] = [];

      querySnapshot.forEach((doc) => {
        const data = doc.data();
        ruches.push({
          id: doc.id,
          idRucher: data.idRucher,
          nom: data.nom,
          position: data.position,
          enService: data.enService,
          dateInstallation: data.dateInstallation?.toDate() || new Date(),
          dateCreation: data.dateCreation?.toDate(),
          dateMiseAJour: data.dateMiseAJour?.toDate(),
          idApiculteur: data.idApiculteur,
          actif: data.actif ?? true,
          nombreCadres: data.nombreCadres,
          typeRuche: data.typeRuche,
          notes: data.notes
        });
      });

      return ruches;
    } catch (error) {
      console.error('Erreur lors de la récupération des ruches:', error);
      throw new Error('Impossible de récupérer les ruches.');
    }
  }

  /**
   * Récupère toutes les ruches de l'utilisateur connecté
   */
  static async obtenirRuchesUtilisateur(): Promise<RucheAvecRucher[]> {
    try {
      const currentUser = auth.currentUser;
      if (!currentUser) {
        throw new Error('Utilisateur non connecté');
      }

      const q = query(
        collection(db, this.COLLECTION_NAME),
        where('idApiculteur', '==', currentUser.uid),
        where('actif', '==', true),
        orderBy('dateCreation', 'desc')
      );

      const querySnapshot = await getDocs(q);
      const ruchesAvecRucher: RucheAvecRucher[] = [];

      // Récupérer les informations des ruchers
      for (const docSnapshot of querySnapshot.docs) {
        const data = docSnapshot.data();
        const ruche: RucheAvecRucher = {
          id: docSnapshot.id,
          idRucher: data.idRucher,
          nom: data.nom,
          position: data.position,
          enService: data.enService,
          dateInstallation: data.dateInstallation?.toDate() || new Date(),
          dateCreation: data.dateCreation?.toDate(),
          dateMiseAJour: data.dateMiseAJour?.toDate(),
          idApiculteur: data.idApiculteur,
          actif: data.actif ?? true,
          nombreCadres: data.nombreCadres,
          typeRuche: data.typeRuche,
          notes: data.notes
        };

        // Récupérer les infos du rucher
        try {
          const rucherDoc = await getDoc(doc(db, this.RUCHERS_COLLECTION, data.idRucher));
          if (rucherDoc.exists()) {
            const rucherData = rucherDoc.data();
            ruche.rucherNom = rucherData.nom;
            ruche.rucherAdresse = rucherData.adresse;
          }
        } catch (error) {
          console.warn('Impossible de récupérer les infos du rucher:', data.idRucher);
        }

        ruchesAvecRucher.push(ruche);
      }

      return ruchesAvecRucher;
    } catch (error) {
      console.error('Erreur lors de la récupération des ruches:', error);
      throw new Error('Impossible de récupérer les ruches.');
    }
  }

  /**
   * Récupère une ruche par son ID
   */
  static async obtenirRucheParId(id: string): Promise<Ruche | null> {
    try {
      const currentUser = auth.currentUser;
      if (!currentUser) {
        throw new Error('Utilisateur non connecté');
      }

      const docRef = doc(db, this.COLLECTION_NAME, id);
      const docSnap = await getDoc(docRef);

      if (docSnap.exists()) {
        const data = docSnap.data();
        
        // Vérifier que la ruche appartient à l'utilisateur
        if (data.idApiculteur !== currentUser.uid) {
          throw new Error('Accès non autorisé à cette ruche');
        }

        return {
          id: docSnap.id,
          idRucher: data.idRucher,
          nom: data.nom,
          position: data.position,
          enService: data.enService,
          dateInstallation: data.dateInstallation?.toDate() || new Date(),
          dateCreation: data.dateCreation?.toDate(),
          dateMiseAJour: data.dateMiseAJour?.toDate(),
          idApiculteur: data.idApiculteur,
          actif: data.actif ?? true,
          nombreCadres: data.nombreCadres,
          typeRuche: data.typeRuche,
          notes: data.notes
        };
      }

      return null;
    } catch (error) {
      console.error('Erreur lors de la récupération de la ruche:', error);
      throw new Error('Impossible de récupérer la ruche.');
    }
  }

  /**
   * Met à jour une ruche
   */
  static async mettreAJourRuche(id: string, donnees: Partial<Ruche>): Promise<void> {
    try {
      const currentUser = auth.currentUser;
      if (!currentUser) {
        throw new Error('Utilisateur non connecté');
      }

      await runTransaction(db, async (transaction) => {
        const rucheRef = doc(db, this.COLLECTION_NAME, id);
        const rucheDoc = await transaction.get(rucheRef);

        if (!rucheDoc.exists()) {
          throw new Error('Ruche non trouvée');
        }

        const rucheData = rucheDoc.data();
        if (rucheData.idApiculteur !== currentUser.uid) {
          throw new Error('Accès non autorisé à cette ruche');
        }

        // Préparer les données de mise à jour
        const updateData: any = { ...donnees };
        delete updateData.id;
        delete updateData.dateCreation;
        delete updateData.idApiculteur;
        
        // Convertir les dates
        if (updateData.dateInstallation instanceof Date) {
          updateData.dateInstallation = Timestamp.fromDate(updateData.dateInstallation);
        }
        
        updateData.dateMiseAJour = Timestamp.now();

        transaction.update(rucheRef, updateData);
      });

      console.log('Ruche mise à jour avec succès');
    } catch (error) {
      console.error('Erreur lors de la mise à jour de la ruche:', error);
      throw new Error('Impossible de mettre à jour la ruche.');
    }
  }

  /**
   * Supprime une ruche (suppression logique)
   */
  static async supprimerRuche(id: string): Promise<void> {
    try {
      const currentUser = auth.currentUser;
      if (!currentUser) {
        throw new Error('Utilisateur non connecté');
      }

      await runTransaction(db, async (transaction) => {
        const rucheRef = doc(db, this.COLLECTION_NAME, id);
        const rucheDoc = await transaction.get(rucheRef);

        if (!rucheDoc.exists()) {
          throw new Error('Ruche non trouvée');
        }

        const rucheData = rucheDoc.data();
        if (rucheData.idApiculteur !== currentUser.uid) {
          throw new Error('Accès non autorisé à cette ruche');
        }

        // Suppression logique
        transaction.update(rucheRef, {
          actif: false,
          dateMiseAJour: Timestamp.now()
        });

        // Mettre à jour le nombre de ruches dans le rucher
        const rucherRef = doc(db, this.RUCHERS_COLLECTION, rucheData.idRucher);
        const rucherDoc = await transaction.get(rucherRef);
        
        if (rucherDoc.exists()) {
          const rucherData = rucherDoc.data();
          const currentNombreRuches = rucherData.nombreRuches || 0;
          transaction.update(rucherRef, {
            nombreRuches: Math.max(0, currentNombreRuches - 1)
          });
        }
      });

      console.log('Ruche supprimée avec succès');
    } catch (error) {
      console.error('Erreur lors de la suppression de la ruche:', error);
      throw new Error('Impossible de supprimer la ruche.');
    }
  }

  /**
   * Écoute les changements des ruches d'un rucher en temps réel
   */
  static ecouterRuchesParRucher(
    idRucher: string,
    callback: (ruches: Ruche[]) => void
  ): () => void {
    const currentUser = auth.currentUser;
    if (!currentUser) {
      throw new Error('Utilisateur non connecté');
    }

    const q = query(
      collection(db, this.COLLECTION_NAME),
      where('idRucher', '==', idRucher),
      where('idApiculteur', '==', currentUser.uid),
      where('actif', '==', true),
      orderBy('position')
    );

    return onSnapshot(q, (querySnapshot) => {
      const ruches: Ruche[] = [];
      querySnapshot.forEach((doc) => {
        const data = doc.data();
        ruches.push({
          id: doc.id,
          idRucher: data.idRucher,
          nom: data.nom,
          position: data.position,
          enService: data.enService,
          dateInstallation: data.dateInstallation?.toDate() || new Date(),
          dateCreation: data.dateCreation?.toDate(),
          dateMiseAJour: data.dateMiseAJour?.toDate(),
          idApiculteur: data.idApiculteur,
          actif: data.actif ?? true,
          nombreCadres: data.nombreCadres,
          typeRuche: data.typeRuche,
          notes: data.notes
        });
      });
      callback(ruches);
    });
  }

  /**
   * Écoute les changements de toutes les ruches de l'utilisateur en temps réel
   */
  static ecouterRuchesUtilisateur(
    callback: (ruches: RucheAvecRucher[]) => void
  ): () => void {
    const currentUser = auth.currentUser;
    if (!currentUser) {
      throw new Error('Utilisateur non connecté');
    }

    const q = query(
      collection(db, this.COLLECTION_NAME),
      where('idApiculteur', '==', currentUser.uid),
      where('actif', '==', true),
      orderBy('dateCreation', 'desc')
    );

    return onSnapshot(q, async (querySnapshot) => {
      const ruchesAvecRucher: RucheAvecRucher[] = [];
      
      for (const docSnapshot of querySnapshot.docs) {
        const data = docSnapshot.data();
        const ruche: RucheAvecRucher = {
          id: docSnapshot.id,
          idRucher: data.idRucher,
          nom: data.nom,
          position: data.position,
          enService: data.enService,
          dateInstallation: data.dateInstallation?.toDate() || new Date(),
          dateCreation: data.dateCreation?.toDate(),
          dateMiseAJour: data.dateMiseAJour?.toDate(),
          idApiculteur: data.idApiculteur,
          actif: data.actif ?? true,
          nombreCadres: data.nombreCadres,
          typeRuche: data.typeRuche,
          notes: data.notes
        };

        // Récupérer les infos du rucher (de manière asynchrone)
        try {
          const rucherDoc = await getDoc(doc(db, this.RUCHERS_COLLECTION, data.idRucher));
          if (rucherDoc.exists()) {
            const rucherData = rucherDoc.data();
            ruche.rucherNom = rucherData.nom;
            ruche.rucherAdresse = rucherData.adresse;
          }
        } catch (error) {
          console.warn('Impossible de récupérer les infos du rucher:', data.idRucher);
        }

        ruchesAvecRucher.push(ruche);
      }
      
      callback(ruchesAvecRucher);
    });
  }

  /**
   * Récupère les mesures des 7 derniers jours d'une ruche depuis Firebase
   */
  static async obtenirMesures7DerniersJours(rucheId: string): Promise<DonneesCapteur[]> {
    return await DonneesCapteursService.getMesures7DerniersJours(rucheId);
  }





  /**
   * Crée des données de test pour une ruche (utilise Firebase)
   */
  static async creerDonneesTest(rucheId: string, nombreJours: number = 10, mesuresParJour: number = 8): Promise<number> {
    return await DonneesCapteursService.creerDonneesTest(rucheId, nombreJours, mesuresParJour);
  }
} 